typescript/cli/src/context/altvm.ts

```diff
@@ -95,43 +150,46 @@ async function loadPrivateKey(
   return keyByProtocol[protocol]!;
 }
 
-export async function createAltVMSigners(
+export function createAltVMSignerGetter(
   metadataManager: ChainMetadataManager,
-  chains: string[],
   keyByProtocol: SignerKeyProtocolMap,
   strategyConfig: Partial<ExtendedChainSubmissionStrategy>,
-) {
+  protocolProviders: ProtocolProviderMap,
+): AltVMSignerGetter {
   const signers: ChainMap<AltVM.ISigner<AnnotatedTx, TxReceipt>> = {};
-  for (const chain of chains) {
-    const metadata = metadataManager.getChainMetadata(chain);
 
-    if (!hasProtocol(metadata.protocol)) {
-      continue;
+  return async (chain: ChainName) => {
+    if (signers[chain]) {
+      return signers[chain]!;
+    }
+
+    const metadata = metadataManager.getChainMetadata(chain);
+    const protocol = metadata.protocol;
+    const providerFactory = protocolProviders[protocol];
+    if (!providerFactory) {
+      throw new Error(`AltVM signer not initialised for protocol ${protocol}`);
     }
 
     const signerConfig = {
       privateKey: await loadPrivateKey(
```

Comment:
We should later change this to request keys up-front. User interaction shouldn't come on-demand, at random times.